File: flutter/test/engage_matrix_test.dart
import 'package:digia_engage/api/internal/action/engage_action.dart';
import 'package:digia_engage/api/internal/event/engage_matrix.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('nudge matrix properties', () {
    test('clicked props derive cta_position, action_type and action_url '
        'from a primary url button', () {
      final props = nudgeClickedProperties(
        label: 'Shop now',
        isPrimary: true,
        actions: const [OpenUrlAction('https://shop.example.com')],
        timeToActionMs: 1500,
      );

      expect(props['element_id'], equals('cta_primary'));
      expect(props['cta_label'], equals('Shop now'));
      expect(props['cta_position'], equals('primary'));
      expect(props['action_type'], equals('url'));
      expect(props['action_url'], equals('https://shop.example.com'));
      expect(props['time_to_action_ms'], equals(1500));
    });

    test('clicked props map a deeplink secondary button without a url for '
        'dismiss/share actions', () {
      final deeplink = nudgeClickedProperties(
        label: 'Open',
        isPrimary: false,
        actions: const [OpenDeeplinkAction('myapp://route/42')],
      );
      expect(deeplink['element_id'], equals('cta_secondary'));
      expect(deeplink['cta_position'], equals('secondary'));
      expect(deeplink['action_type'], equals('deeplink'));
      expect(deeplink['action_url'], equals('myapp://route/42'));
      expect(deeplink.containsKey('time_to_action_ms'), isFalse);

      final dismiss = nudgeClickedProperties(
        label: 'Maybe later',
        isPrimary: false,
        actions: const [HideAction()],
      );
      expect(dismiss['action_type'], equals('dismiss'));
      expect(dismiss.containsKey('action_url'), isFalse);

      final share = nudgeClickedProperties(
        label: 'Share',
        isPrimary: true,
        actions: const [ShareAction('hi')],
      );
      expect(share['action_type'], equals('custom'));
      expect(share.containsKey('action_url'), isFalse);
    });

    test('viewed props carry display_style and optional context only when set',
        () {
      final full = nudgeViewedProperties(
        displayStyle: 'bottom_sheet',
        screenName: 'home',
        triggerType: 'event',
        triggerEvent: 'add_to_cart',
      );
      expect(full, equals({
        'display_style': 'bottom_sheet',
        'screen_name': 'home',
        'trigger_type': 'event',
        'trigger_event': 'add_to_cart',
      }));

      final minimal = nudgeViewedProperties(displayStyle: 'dialog');
      expect(minimal, equals({'display_style': 'dialog'}));
    });

    test('dismissed props carry reason and timing only when set', () {
      expect(
        nudgeDismissedProperties(
          dismissReason: 'backdrop',
          timeToDismissMs: 4200,
        ),
        equals({'dismiss_reason': 'backdrop', 'time_to_dismiss_ms': 4200}),
      );
      expect(nudgeDismissedProperties(), equals(const {}));
    });
  });

  group('inline matrix properties', () {
    test('experience-viewed props stamp the container display_style and total',
        () {
      expect(
        inlineViewedProperties(
          displayStyle: 'carousel',
          itemTotal: 3,
          screenName: 'home',
        ),
        equals({
          'display_style': 'carousel',
          'item_total': 3,
          'screen_name': 'home',
        }),
      );
      // story rail viewed, no screen context known
      expect(
        inlineViewedProperties(displayStyle: 'story', itemTotal: 5),
        equals({'display_style': 'story', 'item_total': 5}),
      );
    });

    test('step-viewed props carry the 0-based index, total and optional id', () {
      expect(
        inlineStepProperties(
          displayStyle: 'carousel',
          itemIndex: 1,
          itemTotal: 3,
          itemId: 'slide_b',
        ),
        equals({
          'display_style': 'carousel',
          'item_index': 1,
          'item_total': 3,
          'item_id': 'slide_b',
        }),
      );
      // no server id → item_id omitted
      expect(
        inlineStepProperties(displayStyle: 'story', itemIndex: 0, itemTotal: 4),
        equals({'display_style': 'story', 'item_index': 0, 'item_total': 4}),
      );
    });

    test('step-clicked props add the resolved action target', () {
      expect(
        inlineStepProperties(
          displayStyle: 'carousel',
          itemIndex: 2,
          itemTotal: 3,
          actionType: 'deeplink',
          actionUrl: 'myapp://pdp/9',
        ),
        equals({
          'display_style': 'carousel',
          'item_index': 2,
          'item_total': 3,
          'action_type': 'deeplink',
          'action_url': 'myapp://pdp/9',
        }),
      );
    });

    test('completed props carry total and optional time_to_complete_ms', () {
      expect(
        inlineCompletedProperties(
          displayStyle: 'story',
          itemTotal: 4,
          timeToCompleteMs: 12000,
        ),
        equals({
          'display_style': 'story',
          'item_total': 4,
          'time_to_complete_ms': 12000,
        }),
      );
      expect(
        inlineCompletedProperties(displayStyle: 'story', itemTotal: 4),
        equals({'display_style': 'story', 'item_total': 4}),
      );
    });
  });
}

File: android/digia-engage/src/test/java/com/digia/engage/internal/analytics/EngageMatrixTest.kt
package com.digia.engage.internal.analytics

import com.digia.engage.internal.model.DismissAction
import com.digia.engage.internal.model.OpenDeeplinkAction
import com.digia.engage.internal.model.OpenUrlAction
import com.digia.engage.internal.model.ShareAction
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/** Pure-builder tests mirroring the Flutter `engage_matrix_test.dart`. */
class EngageMatrixTest {

    @Test
    fun `nudgeClicked synthesises element_id and resolves first action`() {
        val props = EngageMatrix.nudgeClicked(
            label = "Shop now",
            isPrimary = true,
            actions = listOf(OpenUrlAction("https://x.test")),
            timeToActionMs = 1200L,
        )

        assertEquals("cta_primary", props["element_id"])
        assertEquals("Shop now", props["cta_label"])
        assertEquals("primary", props["cta_position"])
        assertEquals("url", props["action_type"])
        assertEquals("https://x.test", props["action_url"])
        assertEquals(1200L, props["time_to_action_ms"])
    }

    @Test
    fun `nudgeClicked secondary without actions omits action fields`() {
        val props = EngageMatrix.nudgeClicked(
            label = "Later",
            isPrimary = false,
            actions = emptyList(),
        )

        assertEquals("cta_secondary", props["element_id"])
        assertEquals("secondary", props["cta_position"])
        assertFalse(props.containsKey("action_type"))
        assertFalse(props.containsKey("action_url"))
        assertFalse(props.containsKey("time_to_action_ms"))
    }

    @Test
    fun `dismiss and share actions map to tokens without urls`() {
        assertEquals("dismiss", EngageMatrix.actionTypeToken(DismissAction))
        assertEquals("custom", EngageMatrix.actionTypeToken(ShareAction("hi")))
        assertEquals("deeplink", EngageMatrix.actionTypeToken(OpenDeeplinkAction("app://x")))
        assertEquals(null, EngageMatrix.actionUrlOf(DismissAction))
        assertEquals("app://x", EngageMatrix.actionUrlOf(OpenDeeplinkAction("app://x")))
    }

    @Test
    fun `nudgeViewed includes context only when known`() {
        val bare = EngageMatrix.nudgeViewed(displayStyle = "dialog")
        assertEquals("dialog", bare["display_style"])
        assertFalse(bare.containsKey("screen_name"))
        assertFalse(bare.containsKey("trigger_type"))

        val full = EngageMatrix.nudgeViewed(
            displayStyle = "bottom_sheet",
            screenName = "Home",
            triggerType = "event",
            triggerEvent = "add_to_cart",
        )
        assertEquals("Home", full["screen_name"])
        assertEquals("event", full["trigger_type"])
        assertEquals("add_to_cart", full["trigger_event"])
    }

    @Test
    fun `containerViewed carries display style and item total`() {
        val props = EngageMatrix.containerViewed(displayStyle = "carousel", itemTotal = 3)
        assertEquals("carousel", props["display_style"])
        assertEquals(3, props["item_total"])
    }

    @Test
    fun `step carries index, total and optional click action`() {
        val viewed = EngageMatrix.step(displayStyle = "carousel", itemIndex = 0, itemTotal = 4)
        assertEquals(0, viewed["item_index"])
        assertEquals(4, viewed["item_total"])
        assertFalse(viewed.containsKey("action_type"))

        val clicked = EngageMatrix.step(
            displayStyle = "carousel",
            itemIndex = 2,
            itemTotal = 4,
            actionType = "deeplink",
            actionUrl = "app://pdp/5",
        )
        assertEquals("deeplink", clicked["action_type"])
        assertEquals("app://pdp/5", clicked["action_url"])
    }

    @Test
    fun `completed includes time only when known`() {
        val bare = EngageMatrix.completed(displayStyle = "story", itemTotal = 5)
        assertFalse(bare.containsKey("time_to_complete_ms"))

        val timed = EngageMatrix.completed(displayStyle = "story", itemTotal = 5, timeToCompleteMs = 8000L)
        assertEquals(8000L, timed["time_to_complete_ms"])
    }

    @Test
    fun `question carries index, total and server identity`() {
        val props = EngageMatrix.question(
            questionIndex = 1,
            questionTotal = 6,
            questionId = "q-nps",
            questionType = "nps",
        )
        assertEquals(1, props["item_index"])
        assertEquals(6, props["item_total"])
        assertEquals("q-nps", props["question_id"])
        assertEquals("nps", props["question_type"])
        assertTrue(props.containsKey("question_type"))
    }
}
